third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp

```diff
@@ -515,18 +457,77 @@ struct TritonAMDGPUInsertInstructionSchedHints
     MLIRContext *ctx = &getContext();
     ModuleOp mod = getOperation();
 
-    mod.walk([this, ctx](scf::ForOp forOp) {
-      // Note, instruction schedule barriers are inserted only in the case of
-      // a single `tt.dot` op in a `scf::ForOp` scope in the current
-      // implementation.
-      if (auto dotOp = getSingleDotOpIfExists(forOp)) {
-        OpBuilder rewriter(ctx);
-        rewriter.setInsertionPointAfter(dotOp);
-        rewriter.create<triton::amdgpu::InstructionSchedHint>(dotOp->getLoc());
+    constexpr llvm::StringLiteral schedRegionIdxAttrName =
+        SchedRegionIdxAttr::getMnemonic();
+    static size_t globalCounter = 0;
+    mod.walk([&](triton::FuncOp funcOp) {
+      auto forOps = AMD::getLeafForOps(funcOp);
+      for (auto forOp : forOps) {
+        forOp->walk([&](triton::DotOp dotOp) {
+          OpBuilder rewriter(ctx);
+          rewriter.setInsertionPointAfter(dotOp);
+          auto hintOp = rewriter.create<InstructionSchedHint>(dotOp->getLoc());
+
+          auto regionIdx = SchedRegionIdxAttr::get(ctx, globalCounter);
+          dotOp->setAttr(schedRegionIdxAttrName, regionIdx);
+          hintOp->setAttr(schedRegionIdxAttrName, regionIdx);
+          ++globalCounter;
+        });
       }
     });
   }
 };
+
+struct TritonAMDGPUInsertInstructionSchedGuards
+    : public triton::impl::TritonAMDGPUInsertInstructionSchedGuardsBase<
+          TritonAMDGPUInsertInstructionSchedGuards> {
+
+  void runOnOperation() override {
+    MLIRContext *ctx = &getContext();
+    ModuleOp mod = getOperation();
+
+    llvm::SmallVector<scf::ForOp> forOps;
+    mod.walk([&](scf::ForOp forOp) {
+      auto result = forOp->walk(
+          [](InstructionSchedHint) { return WalkResult::interrupt(); });
+      if (result.wasInterrupted())
+        forOps.push_back(forOp);
+    });
+
+    for (auto forOp : forOps) {
+      OpBuilder rewriter(ctx);
+      int32_t prevSchedRegionIdx = -1;
+      auto body = forOp.getBody();
+      for (auto op = body->begin(); op != body->end(); ++op) {
+        if (auto currRegionIdx = op->getAttrOfType<SchedRegionIdxAttr>(
+                SchedRegionIdxAttr::getMnemonic())) {
+          if (currRegionIdx.getValue() != prevSchedRegionIdx) {
+            if (op == body->begin()) {
+              rewriter.setInsertionPointToStart(body);
+            } else {
+              rewriter.setInsertionPointAfter(&(*(std::prev(op))));
+            }
+            createSchedBarrier(rewriter, op->getLoc(),
+                               mlir::amdgpu::sched_barrier_opt_enum::none);
+            prevSchedRegionIdx = currRegionIdx.getValue();
+          }
+        }
+      }
+      prevSchedRegionIdx = -1;
+      for (auto op = body->rbegin(); op != body->rend(); ++op) {
```

Comment:
Why a second pass in the reverse order? Are you inserting the schedule barrier just after the last case? If so you can remember what is the last case in the previous pass and do some patch up just here?